src/array.rs

```diff
@@ -0,0 +1,456 @@
+        bf16::from_bits(val)
+    }
+
+    fn array_data(array: &Array) -> *const Self {
+        unsafe { mlx_sys::mlx_array_data_bfloat16(array.ptr) as *const Self }
+    }
+}
+
+impl Sealed for c64 {}
+
+impl ArrayElement for c64 {
+    const DTYPE: Dtype = Dtype::Complex64;
+
+    fn scalar_array_item(array: &Array) -> Self {
+        let bindgen_c64 = unsafe { mlx_sys::mlx_array_item_complex64(array.ptr) };
+
+        Self {
+            re: bindgen_c64.re,
+            im: bindgen_c64.im,
+        }
+    }
+
+    fn array_data(array: &Array) -> *const Self {
+        // c64 has the same memory layout as __BindgenComplex<f32>
+        unsafe { mlx_sys::mlx_array_data_complex64(array.ptr) as *const Self }
+    }
+}
+
+pub struct Array {
+    ptr: mlx_array,
+}
+
+impl std::fmt::Debug for Array {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let description = crate::utils::mlx_describe(self.ptr as *mut c_void)
+            .unwrap_or_else(|| "Array".to_string());
+        write!(f, "{:?}", description)
+    }
+}
+
+// TODO: Clone should probably NOT be implemented because the underlying pointer is atomically
+// reference counted but not guarded by a mutex.
+
+impl Drop for Array {
+    fn drop(&mut self) {
+        // TODO: check memory leak with some tool?
+
+        // Decrease the reference count
+        unsafe { mlx_sys::mlx_free(self.ptr as *mut c_void) };
+    }
+}
+
+impl Array {
+    /// Create a new array from an existing mlx_array pointer.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure the reference count of the array is properly incremented with
+    /// `mlx_sys::mlx_retain`.
+    pub unsafe fn from_ptr(ptr: mlx_array) -> Array {
+        Array { ptr }
+    }
+
+    // TODO: should this be unsafe?
+    pub fn as_ptr(&self) -> mlx_array {
+        self.ptr
+    }
+
+    /// New array from a bool scalar.
+    pub fn from_bool(val: bool) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_bool(val) };
+        Array { ptr }
+    }
+
+    /// New array from a int scalar.
+    pub fn from_int(val: i32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_int(val) };
+        Array { ptr }
+    }
+
+    /// New array from a float scalar.
+    pub fn from_float(val: f32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_float(val) };
+        Array { ptr }
+    }
+
+    /// New array from a complex scalar.
+    pub fn from_complex(val: c64) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_complex(val.re, val.im) };
+        Array { ptr }
+    }
+
+    /// New array from existing buffer.
+    ///
+    /// # Parameters
+    ///
+    /// - `data`: A buffer which will be copied.
+    /// - `shape`: Shape of the array.
+    /// - `dim`: Number of dimensions (size of shape).
+    pub fn from_data<T: ArrayElement>(data: &[T], shape: &[i32], dim: i32) -> Self {
+        let ptr = unsafe {
```

Comment:
We'll want to add a precondition here to validate the shape and data length. something like:

```rust
let total = shape.iter().fold(1, |acc, &x| acc * x);
assert!(total == data.len(), "shape {shape} total {total} != {count} (actual)")
```

Reply:
I think mlx c++ performs this check already. Need to double check. But panicking on the c++ side would give an fatal error on the rust side, which we will need to deal with at some point

Reply:
I'd say where we can catch them on the Rust side, let's just do it

Reply:
This function has been renamed to `from_slice`, which now checks whether the shape and data length match